### PR TITLE
Lindsey Deuel - Pipes - MediaRanker

### PR DESCRIPTION
**Pull request by VirtualLindsey**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a method to pull the number of works a user has voted on
Describe how you approached testing that model method. What edge cases did you come up with? | I checked for valid users and invalid users
Describe an edge case test you wrote for a controller | Checking to see if the user exists for the show
What are `session` and `flash`? What is the difference between them? | Session is used to track who is and isn't logged in. Flash is used to display messages between redirects
Describe a controller filter you wrote. |
What was one thing that you gained more clarity on through this assignment? | Using sessions and flashes successfully
What is the Heroku URL of your deployed application |
Do you have any recommendations on how we could improve this project for the next cohort? | More examples would be helpful.

### CODE REVIEW

File: app/controllers/works_controller.rb [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  def create
+    if params[:work][:title] == ""
+      flash[:failed] = "title can't be blank"
+      redirect_to new_work_path
+    else
+
+      @work = Work.new(title: params[:work][:title],
+                      description: params[:work][:description],
+                      category: params[:work][:category],
+                      creator: params[:work][:creator],
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb]: You should be using strong params here to filter input from the user.

File: app/controllers/works_controller.rb [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  end
+
+  def new
+    @work = Work.new
+  end
+
+  def create
+    if params[:work][:title] == ""
+      flash[:failed] = "title can't be blank"
+      redirect_to new_work_path
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb]: There are more ways for creating a work to fail than just the title being blank. Instead of checking things explicitly here, you should rely on your model validations to tell you whether or not the generated model is valid, then report their automatically-generated error messages to the user. Something like:

```ruby
def create
  # Using strong params here as an initial filter on user input
  @work = Work.new(work_params)
  # Validations will be run automatically when .save is called. The
  # return value will be whether or not the validations passed.
  if @work.save
    flash[:success] = "Created new work"
    redirect_to work_path(@work)
  else
    flash.now[:failure] = "Could not create work"
    # If the work failed to save, @work.errors.messages will be populated with error messages
    flash.now[:details] = @work.errors.messages
    render :new
  end
end
```

File: app/controllers/home_controller.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class HomeController < ApplicationController
+  def index
+    @top_work_id = Vote.limit(1).group(:work_id).count
+    @top_work = Work.find(@top_work_id.keys[0])
+    @top_work_votes = Vote.where(work_id: @top_work_id.keys[0])
```
Comment (droberts-sea) [inline app/controllers/home_controller.rb]: This fails if there are no votes!

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+      if user
+        flash[:notice] = 'Logged in as existing user'
+        session[:user_id] = user.id
+        redirect_to root_path, notice: "Successfully logged in as existing user #{user.name}"
+      else
+        @new_user = User.new
+        @new_user.joined_on = Date.today
+        @new_user.name = params[:name]
+        @new_user.save
```
Comment (droberts-sea) [inline app/controllers/sessions_controller.rb]: You should be checking the return value of `@new_user.save` here. As is, saving could fail and the user wouldn't be informed why.

File: config/routes.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  get '/works/new', to: 'works#new', as: 'new_work'
+  post '/works', to: 'works#create'
+  post '/works/:id/upvote', to: 'works#upvote', as: 'upvote_work'
+  get '/works/:id', to: 'works#show', as: 'work'
+  get '/works/:id/edit', to: 'works#edit', as: 'edit_work'
+  patch '/works/:id', to: 'works#update'
+  delete '/works/:id', to: 'works#destroy'
+  root 'home#index'
+#  root 'works#index'
+  resources :users, :books, :movies, :albums, :sessions
```
Comment (droberts-sea) [inline config/routes.rb]: Observations here:

1. The manually specified routes above look good, although it might be a good choice to use `resources` for works, since you've got all seven of the standard restful routes.
1. If you're using `resources` for `:users`, you should add an `only:` argument so that you're only generating routes you need.
1. The `:books`, `:movies`, `:albums` and `:sessions` routes generated via `resources` here aren't used at all in your site! Make sure you're careful about keeping your routes clean.

File: app/views/layouts/_header.html.erb [hunk 6]
```diff
@@ -0,0 +1,48 @@
+<header class='header row'>
+
+    <h2 class = 'columns small-12'>
```
Comment (droberts-sea) [inline app/views/layouts/_header.html.erb]: Since this code is rendered as a partial on every page, why not add it to `application.html.erb` and save yourself having to remember to `render partial`.

File: app/views/works/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+<main class='row'>
+  <h3> List of works </h3>
+
+  <section>
+    <h4>
+      Albums
+    </h4>
+    <table class ='table'>
+      <thead>
```
Comment (droberts-sea) [inline app/views/works/index.html.erb]: Since this table is repeated several times, it would be a good candidate to move into a partial.

File: test/models/work_test.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+2 tests for validation
+2 tests on scope
+=end
+
+describe Work do
+  it "must be invalid" do
+    work = Work.new
+    result = work.valid?
+    result.must_equal false
+    #value(user).must_be :valid?
```
Comment (droberts-sea) [inline test/models/work_test.rb]: As we discussed in class, you should have a separate test for each validation, to make sure that works that are invalid in this way do actually cause it to fail.

In each test, you should check that `errors.messages` contains the field that you expected to be invalid, to ensure that the validation failure wasn't coincidental.

File: test/controllers/user_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+    it 'test votes_for_work' do
+      work = Work.find(1)
+      t = ApplicationHelper.votes_for_work(work)
+      t.must_be_kind_of Integer
+    end
+
+    it 'tests top_10' do
+      result = ApplicationHelper.top_10('album')
+      result.must_be_kind_of Array
```
Comment (droberts-sea) [inline test/controllers/user_controller_test.rb]: Checking that top_10 returns an array is a good start, but I would like to see a little more in-depth coverage here. For example:
- Are works returned in order by number of votes?
- Are all returned works of the correct category?
- What happens if there are zero works with that category?
- What happens if there are more than 10 works?

File: test/controllers/user_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        }
+      }
+      must_respond_with :redirect
+      must_redirect_to root_path
+    end
+  end
+
+  describe 'testing custom functions' do
+    it 'test find_user' do
+      t = ApplicationHelper.find_user(users(:one).id)
```
Comment (droberts-sea) [inline test/controllers/user_controller_test.rb]: I like your dedication to testing these methods. We haven't talked about this in class, but Rails provides a separate folder for view helper tests: `test/helpers`, and it would probably make sense to move these there.

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes - good work
Comprehension questions	| yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | some - see inline comment in the route file
Semantic HTML | yes - good work. See inline comments about HTML structure.
Errors are reported to the user | some - see inline comment
Business logic lives in the models | no - You do a good job of extracting functionality to helper methods, but much of what you have as view helpers in `ApplicationHelper` would be a better fit as model logic. For example, getting the top work by vote count, or the top 10 works for a given category, would both be excellent `self` methods on the `Work` class. Then in your controller or view you could call `Work.top_10(:album)` instead of `ApplicationHelper.top_10(:album)`, which I would argue is more readable.
Models are thoroughly tested, including relations, validations and any custom logic | no - see inline comments. 
Controllers are thoroughly tested | some - many of the tests for `WorksController` are missing
**Wave 1 - Media** |
Splash page shows the three media categories | yes, but fails with no votes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
ActiveRecord relations are used appropriately | no - You don't define any relations between your models! Adding the `belongs_to` and `has_many` rules at the ActiveRecord level gives you a large number of helper methods for free, which would probably make some of your view helpers unnecessary.<br><br>For example, instead of `ApplicationHelper.get_published(vote.work_id)` you could say `vote.work.published`.
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
Individual user pages and the user list are present | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | no
Splash page contains a media spotlight | yes
Media pages contain lists of voting users | yes
**Wave 3 - Styling** |
Foundation is used appropriately | mostly - many components (media lists on the front page, for example) are not responsive
Look and feel is similar to the original | yes
**Overall** |

This site looks good. Other than a couple of small bugs, it seems like you were able to get the features we asked for up and running. I especially appreciate your attention the site's visual styles, since I know you're not particularly fond of working on that.

While I'm happy with the site's functionality, your implementation has me somewhat worried. It seems like there are some major holes in your understanding of the Rails curriculum, particularly around working with models. This is understandable given you missed some time in week 9, but it's still something we need to address. My recommendation is to schedule one or two tutoring sessions this week to make sure these details are solidified. I've spoken with Charles about this, and he should be reaching out to you soon to figure out availability.

It's also important to note that when you miss class, it is on you to figure out what topics you've missed and make sure you understand them, and to ask for help when you don't. There are many resources available (Panopto, textbook pages, open tutoring, etc) but it's up to you to take advantage of them. We have limited visibility into how you're doing, especially during pair projects, and in a situation like this that means we don't know there's a problem until much later than we'd like.

This is definitely something we can work past. The fact that you were able to complete the project without some of the tools we've taught is a good sign, as are the places where you noticed a problem the tools solve and addressed it your own way (like with the view helpers). Shoring up your understanding of the details of Rails should make future projects go much smoother.
